### PR TITLE
#13173 WinAppDriver does not wait long startup an UWP app

### PR DESCRIPTION
**Pull request by Korop**
add create session timeout and waitForCondition to retry the session creation in case of an error
add 'createSessionTimeout' capability, WindowsDriver only

### CODE REVIEW

File: lib/driver.js [hunk 1]
```diff
@@ -17,6 +17,7 @@ class WindowsDriver extends BaseDriver {
     this.jwpProxyAvoid = NO_PROXY;
 
     this.opts.address = opts.address || DEFAULT_WAD_HOST;
+    this.opts.createSessionTimeout = opts.createSessionTimeout || DEFAULT_CREATE_SESSION_TIMEOUT_MS;
```
Comment (mykola-mokhnach) [inline lib/driver.js]: what's the point to import and assign the default value here if we can do the same inside winappdriver module?

File: lib/winappdriver.js [hunk 2]
```diff
@@ -130,8 +132,40 @@ class WinAppDriver extends events.EventEmitter {
   }
 
   async startSession (caps) {
+
+    let createSessionTimeout = this.createSessionTimeout;
+    if (caps.createSessionTimeout) {
+      createSessionTimeout = Number.isInteger(caps.createSessionTimeout) ? caps.createSessionTimeout : parseInt(caps.createSessionTimeout, 10);
```
Comment (mykola-mokhnach) [inline lib/winappdriver.js]: this is not needed. if the capability is defined as isNumber, that it will be automatically cast to the necessary type

File: lib/winappdriver.js [hunk 3]
```diff
@@ -130,8 +132,40 @@ class WinAppDriver extends events.EventEmitter {
...
+    let lastError = null;
+
+    const condFn = async () => {
+      try {
+        retryIteration++;
+        await this.jwproxy.command('/session', 'POST', {desiredCapabilities: caps});
+        return true;
+      } catch (error) {
+        lastError = error;
+        log.error(`Could not start WinAppDriver session error = '${error}', attempt = '${retryIteration}' from '${this.createSessionRetry}'`);
```
Comment (mykola-mokhnach) [inline lib/winappdriver.js]: I'd rather log a warning here

File: lib/winappdriver.js [hunk 4]
```diff
@@ -130,8 +132,40 @@ class WinAppDriver extends events.EventEmitter {
...
+      }
+    };
+
+    try {
+      await waitForCondition(condFn, {
+        waitMs: createSessionTimeout,
+        intervalMs: 500
+      });
+    } catch (timeoutError) {
+      log.error(timeoutError);
```
Comment (mykola-mokhnach) [inline lib/winappdriver.js]: this log is not needed

File: lib/winappdriver.js [hunk 5]
```diff
@@ -130,8 +132,40 @@ class WinAppDriver extends events.EventEmitter {
...
+      if (lastError) {
+        throw (lastError);
+      }
```
Comment (mykola-mokhnach) [inline lib/winappdriver.js]: else throw timeoutError

File: package.json [hunk 6]
```diff
@@ -54,7 +54,7 @@
     "watch": "gulp watch",
     "coverage": "gulp coveralls",
     "precommit-msg": "echo 'Pre-commit checks...' && exit 0",
-    "precommit-test": "REPORTER=dot gulp once",
+    "precommit-test": "env REPORTER=dot bash -c 'gulp once'",
```
Comment (mykola-mokhnach) [inline package.json]: why this change is necessary?

File: README.md [hunk 7]
```diff
@@ -26,6 +26,13 @@ let driver = new WindowsDriver();
 await driver.createSession(defaultCaps);
 ```
 
+## Additional WindowsDriver only capabilities
```
Comment (mykola-mokhnach) [inline README.md]: WindowsDriver-specific capabilities

File: README.md [hunk 8]
```diff
@@ -26,6 +26,13 @@ let driver = new WindowsDriver();
...
+
+|Capability|Description|Values|
+|----------|-----------|------|
+|`createSessionTimeout`|Timeout in milliseconds used to retry `WinAppDriver` start session creation. This is workaround for a long startup duration UWP application error: `Original error: Failed to locate opened application window with appId: TestCompany.my_app4!App, and processId: 8480`. Default value `20000`|e.g., `15000`|
```
Comment (mykola-mokhnach) [inline README.md]: ...used to retry `WinAppDriver` session startup.

Comment (mykola-mokhnach) [inline README.md]: This is the workaround for the long startup times of UWP applications (aka `Failed to locate opened application window with appId: TestCompany.my_app4!App, and processId: 8480`)

File: lib/winappdriver.js [hunk 9]
```diff
@@ -3,17 +3,18 @@ import { JWProxy } from 'appium-base-driver';
...
-import { retryInterval } from 'asyncbox';
+import { retryInterval, waitForCondition } from 'asyncbox';
 import cp from 'child_process';
 import B from 'bluebird';
 
 const REQUIRED_PARAMS = []; // XXYD 1-5-2018: removed app from required params because you can alternatively use appTopLevelWindow
 const DEFAULT_WAD_HOST = '127.0.0.1';
-const DEFAULT_WAD_PORT = 4724; //  should be non-4723 to avoid conflict on the same box
+const DEFAULT_WAD_PORT = 4724; // should be non-4723 to avoid conflict on the same box;
+const DEFAULT_CREATE_SESSION_TIMEOUT_MS = 20000; // retry start session creation during the timeout in milliseconds
```
Comment (mykola-mokhnach) [inline lib/winappdriver.js]: how long it was before the change?

File: package.json (hunk 6 above)
Comment (Korop) [reply package.json]: on my PC (Windows10) I have permanent error when do commit. precommit hook:
appium-windows-driver> git.exe commit --file=C:/Users/autotest/AppData/Local/Temp/smartgit-675290320427002764tmp/commit-12634442983062762856.tmp -o -- package.json
'Pre-commit checks...' 
'REPORTER' is not recognized as an internal or external command,
operable program or batch file.
pre-commit: 
pre-commit: We've failed to pass the specified git pre-commit hooks as the `precommit-test`
pre-commit: hook returned an exit code (1). If you're feeling adventurous you can
pre-commit: skip the git pre-commit hooks by adding the following flags to your commit:
pre-commit: 
pre-commit:   git commit -n (or --no-verify)
pre-commit: 
pre-commit: This is ill-advised since the commit is broken.
pre-commit:

Comment (mykola-mokhnach) [reply package.json]: @imurchie FYI

File: lib/winappdriver.js [hunk 10]
```diff
@@ -202,5 +236,5 @@ WinAppDriver.STATE_STARTING = 'starting';
 WinAppDriver.STATE_ONLINE = 'online';
 WinAppDriver.STATE_STOPPING = 'stopping';
 
-export { WinAppDriver, DEFAULT_WAD_HOST, DEFAULT_WAD_PORT};
+export { WinAppDriver, DEFAULT_WAD_HOST, DEFAULT_WAD_PORT, DEFAULT_CREATE_SESSION_TIMEOUT_MS};
```
Comment (mykola-mokhnach) [inline lib/winappdriver.js]: do we use the constant anywhere else except of this module?

File: lib/winappdriver.js (hunk 2 above)
Comment (mykola-mokhnach) [reply lib/winappdriver.js]: why this comment got resolved without an explanation? the original problem is still there

Comment (mykola-mokhnach) [reply lib/winappdriver.js]: Are you sure this is the case? https://github.com/appium/appium-base-driver/blob/69dde698d6ef33f50b8ec7f88ee0fb0ddc86e1bf/lib/basedriver/commands/session.js#L132 should automatically fix capability types. Do you observe the warning message like https://github.com/appium/appium-base-driver/blob/69dde698d6ef33f50b8ec7f88ee0fb0ddc86e1bf/lib/basedriver/commands/session.js#L142 while running your test and the cap is passed as string?

**Comment by licanhua** [comment]
WinAppDriver is introducing 'ms:waitForAppLaunch' flag in the next preview release and it may help with this problem

File: lib/winappdriver.js (hunk 2 above)
Comment (Korop) [reply lib/winappdriver.js]: @mykola-mokhnach you was right, the automatic cast has to be, I have debugged the base-driver code and found that desiredCapConstraints is not really updated by subclass and all windows driver specific constraints are ignored. So I added this.desiredCapConstraints = desiredCapConstraints; to the driver constructor. That fixes the issue with the cast.

**Comment by Korop** [comment]
> WinAppDriver is introducing 'ms:waitForAppLaunch' flag in the next preview release and it may help with this problem

@licanhua 
That will be nice and right place for the feature! But till this happens I prefer to have the workaround.

File: lib/driver.js [hunk 11]
```diff
@@ -12,10 +13,9 @@ const NO_PROXY = [
 class WindowsDriver extends BaseDriver {
   constructor (opts = {}, shouldValidateCaps = true) {
     super(opts, shouldValidateCaps);
+    this.desiredCapConstraints = desiredCapConstraints;
```
Comment (Korop) [inline lib/driver.js]: Fix for the constraints update with driver specific ones

File: lib/winappdriver.js [hunk 12]
```diff
@@ -130,8 +132,37 @@ class WinAppDriver extends events.EventEmitter {
   }
 
   async startSession (caps) {
+
+    let createSessionTimeout = caps.createSessionTimeout || this.createSessionTimeout;
+    log.info(`Start WinAppDriver session during createSessionTimeout = '${createSessionTimeout}' ms.`);
```
Comment (mykola-mokhnach) [inline lib/winappdriver.js]: this log message is not needed

File: lib/winappdriver.js [hunk 13]
```diff
@@ -130,8 +132,37 @@ class WinAppDriver extends events.EventEmitter {
   }
 
   async startSession (caps) {
+
+    let createSessionTimeout = caps.createSessionTimeout || this.createSessionTimeout;
```
Comment (mykola-mokhnach) [inline lib/winappdriver.js]: const

File: lib/winappdriver.js [hunk 14]
```diff
@@ -130,8 +132,37 @@ class WinAppDriver extends events.EventEmitter {
...
+      await waitForCondition(condFn, {
+        waitMs: createSessionTimeout,
+        intervalMs: 500
+      });
+    } catch (timeoutError) {
+      log.debug(timeoutError);
+      if (lastError) {
+        throw (lastError);
+      }
+      throw (timeoutError);
```
Comment (mykola-mokhnach) [inline lib/winappdriver.js]: the native error thrown by `waitForCondition` is not very informative. We could provide a better error message here

File: lib/winappdriver.js [hunk 15]
```diff
@@ -130,8 +132,38 @@ class WinAppDriver extends events.EventEmitter {
   }
 
   async startSession (caps) {
+
+    const createSessionTimeout = caps.createSessionTimeout || this.createSessionTimeout;
+    log.debug(`Start WinAppDriver session during createSessionTimeout = '${createSessionTimeout}' ms.`);
```
Comment (mykola-mokhnach) [inline lib/winappdriver.js]: Then I would rather rephrase the log line to `Starting WinAppDriver session. Will timeout in ... ms`

File: lib/winappdriver.js [hunk 16]
```diff
@@ -130,8 +132,38 @@ class WinAppDriver extends events.EventEmitter {
...
+      await waitForCondition(condFn, {
+        waitMs: createSessionTimeout,
+        intervalMs: 500
+      });
+    } catch (timeoutError) {
+      log.debug(`timeoutError was ${timeoutError}`);
+      if (lastError) {
+        throw (lastError);
+      }
+      throw new Error(`Could not start WinAppDriver session during timeout '${createSessionTimeout}' ms.`);
```
Comment (mykola-mokhnach) [inline lib/winappdriver.js]: Could not start WinAppDriver session within '${createSessionTimeout}' ms.

Comment (Korop) [reply lib/winappdriver.js]: Done

File: lib/winappdriver.js (hunk 15 above)
Comment (Korop) [reply lib/winappdriver.js]: Done

File: lib/winappdriver.js (hunk 16 above)
Comment (Korop) [reply lib/winappdriver.js]: BTW with the current implementation of waitForCondition this Error will be never thrown as it does not check success result on timing out.
